src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -3115,20 +3669,41 @@ private void ImportMkRefAny(int token)
         private void ImportLdToken(int token)
         {
             var ldtokenValue = _methodIL.GetObject(token);
-            StackEntry value;
             if (ldtokenValue is TypeDesc)
             {
+                TypeDesc runtimeTypeHandleTypeDesc = GetWellKnownType(WellKnownType.RuntimeTypeHandle);
                 var typeDesc = (TypeDesc)ldtokenValue;
-                PushLoadExpression(StackValueKind.ByRef, "ldtoken", GetEETypePointerForTypeDesc(typeDesc, false), _compilation.TypeSystemContext.SystemModule.GetKnownType("System", "EETypePtr"));
                 MethodDesc helper = _compilation.TypeSystemContext.GetHelperEntryPoint("LdTokenHelpers", "GetRuntimeTypeHandle");
                 AddMethodReference(helper);
-                HandleCall(helper, helper.Signature);
-                _dependencies.Add(_compilation.NodeFactory.MaximallyConstructableType(typeDesc));
+                var fn = LLVMFunctionForMethod(helper, helper, null/* static method */, false /* not virt */, _constrainedType, helper, out bool hasHiddenParam, out LLVMValueRef dictPtrPtrStore, out LLVMValueRef fatFunctionPtr);
+
+                if (typeDesc.IsRuntimeDeterminedSubtype)
+                {
+                    var hiddenParam = CallGenericHelper(ReadyToRunHelperId.TypeHandle, typeDesc);
+                    var handleRef = LLVM.BuildCall(_builder, fn, new LLVMValueRef[]
+                    {
+                        GetShadowStack(),
+                        hiddenParam
+                    }, "getHelper");
+                    _stack.Push(new LdTokenEntry<TypeDesc>(StackValueKind.ValueType, "ldtoken", typeDesc, handleRef, runtimeTypeHandleTypeDesc));
+                }
+                else
+                {
+                    if (ConstructedEETypeNode.CreationAllowed(typeDesc))
```

Comment:
Do we need this if check? If we call `GetEETypePointerForTypeDesc` with `true` below, it should do the right thing wrt dependency tracking.

Reply:
No, works with it removed.